### DISCUSSION TITLE
Change proposal: Consider Vest for user-facing validations in the Epic Stack

### DISCUSSION THREAD

**Discussion by ealush** [epicweb-dev/epic-stack#605] [Ideas]
Hey 👋
Thanks for making the Epic Stack. I believe it brings actual value to the developer community. I'd like to propose considering [Vest](https://vestjs.dev/) as an alternative or complementary option to Zod for user-facing validations.

**Disclaimer, I am the author of Vest*

Obviously, I don't expect it to be added just because I asked here, and I want to make a proposal PR to make my point. I know reviewing PRs can be a burden so I haven't yet created a PR - I wanted to ask here first before I make a proposal PR.

Here are the broad details that I am going to implement in my PR:

## What

1. Add [Vest validations library](https://vestjs.dev/) alongside Zod or instead of it
2. Add examples for user facing validations, adding examples of interactive validations
3. Update relevant docs to reflect this option.

## Why

I see the value in the Epic Stack as an opinionated tech stack, and that it can greatly boost the productivity of people building their apps.

I believe Vest can improve their productivity, and allow for capabilities currently not possible with Zod, or very hard to implement, along with improving the end-user experience, especially around async validation, as each validation can be async individually and report back immediately as it finishes.

Vest validates each field as the user types, rather than the entire schema, so no need to break the form down into different smaller schemas.

Its declarative unit-testing-like syntax can reduce the complexity of imperative validations like this:

https://github.com/epicweb-dev/epic-stack/blob/1d51912e88cf8df777b6f242d69da970d775004f/app/utils/user-validation.ts#L29-L39
Would have changed to:

```ts
test("confirmPassword", "Passwords do not match", () => {
  enforce(data.confirmPassword).equals(data.password);
});
```

Of course, I'm also open to your feedback and insights on whether this proposed change aligns with the Epic Stack's vision.

Thanks for your time and consideration!

**Comment by kentcdodds** [comment]
Thanks for sharing! I appreciate your interest in the Epic Stack.

I don't think that I'm interested in making this change for a few reasons:

1. It's definitely more verbose (your comparison is an oversimplification because there's a LOT more the schema is doing than comparing the passwords).
2. TypeScript support and inference is critical
3. Integration with Conform is a very important part of the DX as well.

Thanks again for sharing and good luck on your project!

**Reply by ealush** [reply]
Thanks Kent for the response and for considering my proposal.

Have a great weekend!
